Fixes #27174 - REST API-based DNS conflict check

lzap opened the pull request:
Instead relying on Ruby DNS resolv by common DNS module from smart-proxy core.

This took me all day of work, setting up Infoblox DNS, creating IPv6 space etc etc. I tested this end to end using a simple script, I am going to attach it to the PR.

lzap commented:
Script that tests this end to end - I wonder if it makes sense to put it to smart proxy core `extras/` directory or somewhere else:

```ruby
require 'rest-client'
require 'json'
require 'pp'

URL = "http://localhost:8448"
DNS_DOMAIN = "ibx.lan"
DNS_IPV4 = "192.168.202.13"
DNS_IPV6 = "2001:db8::1"
DNS_PTR4 = "13.202.168.192.in-addr.arpa"
DNS_PTR6 = "1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.8.b.d.0.1.0.0.2.ip6.arpa"

def do_get(path)
  RestClient.get("#{URL}#{path}")
end

def do_post(path, body = nil, params = {})
  RestClient.post("#{URL}#{path}", body, params.merge({content_type: :json}))
end

def do_delete(path, params = {})
  RestClient.delete("#{URL}#{path}", params.merge({content_type: :json}))
end

features = JSON.parse(do_get("/v2/features"))

if features["dns"]["state"] == "running"
  fqdn = "end2end.#{DNS_DOMAIN}"
  cname = "alias-end2end.#{DNS_DOMAIN}"
  puts "Performing DNS tests with #{fqdn}"
  begin
    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}")
    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}") # skip (already exists)
    do_post("/dns?type=a&fqdn=#{fqdn}&value=8.8.8.8") rescue RestClient::Conflict # expected
  ensure
    do_delete("/dns/#{fqdn}/A")
  end
  begin
    do_post("/dns?type=aaaa&fqdn=#{fqdn}&value=#{DNS_IPV6}")
    do_post("/dns?type=aaaa&fqdn=#{fqdn}&value=#{DNS_IPV6}") # skip (already exists)
    do_post("/dns?type=aaaa&fqdn=#{fqdn}&value=2001:db8::2") rescue RestClient::Conflict # expected
  ensure
    do_delete("/dns/#{fqdn}/AAAA")
  end
  begin
    do_post("/dns?type=cname&fqdn=#{fqdn}&value=#{cname}")
    do_post("/dns?type=cname&fqdn=#{fqdn}&value=#{cname}") # skip (already exists)
    do_post("/dns?type=cname&fqdn=#{fqdn}&value=a2-#{cname}") rescue RestClient::Conflict # expected
  ensure
    do_delete("/dns/#{fqdn}/CNAME")
  end
  begin
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=#{DNS_PTR4}")
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=#{DNS_PTR4}") # skip (already exists)
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=14.202.168.192.in-addr.arpa") rescue RestClient::Conflict # expected
  ensure
    do_delete("/dns/#{DNS_PTR4}/PTR")
  end
  begin
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=#{DNS_PTR6}")
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=#{DNS_PTR6}") # skip (already exists)
    do_post("/dns?type=ptr&fqdn=#{fqdn}&value=14.202.168.192.in-addr.arpa") rescue RestClient::Conflict # expected
  ensure
    do_delete("/dns/#{DNS_PTR6}/PTR")
  end
end
```

@ekohl

ekohl commented:
In the past I've used https://github.com/theforeman/forklift/blob/master/bats/fb-proxy-dns.bats for this

lzap commented:
This removes need of 79ceedcee3bb3e334cea2f7219a8820ae44dc6dd so I have removed the env. variable from the codebase.

Bats looks good however I need to run this script over and over again during development. I added it here locally into extra/ dir just to be handy for the future.

lzap commented:
Resolved Rubocop issues, any comments @ekohl ?

ekohl commented on lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb:
```diff
@@ -20,9 +20,73 @@ def do_remove(name, type)
+      Infoblox::Cname.find(connection, params)
+    end
+
+    def ib_create_a_record(fqdn, address = nil)
```
When is the address nil? That doesn't sound like a valid use case

lzap replied:
Look at the `record_conflicts_ip` implementation, I could not figure out better way but this is how it works:

* It calls `ib_create_a_record("fqdn.example.com")` first grabbing all records of the specific type. If that's empty it returns -1 (no conflict for sure).
* Then calls `ib_create_a_record("fqdn.example.com", "1.2.3.4")` grabbing all records of the specific type with the specific address. If there is any record that means the record already exists and it returns 0.
* Otherwise it means there are some records with "fqdn.example.com" which are however not set to "1.2.3.4", therefore it returns 1 (conflict).

I know its a little bit lame, but I could not figure out how to do "return me fqdn.example.com A records which don't have value set to 1.2.3.4". This does it, it's just two simple queries both returning just empty array or array with maximum of one record. Alternatively I could do just one query grabbing ALL records of the particular type and fqdn and then iterating them, but possibly this could return many values (Infoblox entries also have lots of fields). I wanted to avoid that.

So back to the concern - all these helper methods can be called with or without 2nd parameter. It's also the reason why I swap params for the PTR record because it's actually swapped by the common code and only last parameter(s) are optional in Ruby.

ekohl replied:
That calls `ib_find_a_record` and calling that without an address makes sense but this is the create method. Creation without an address sounds invalid.

I will also agree that the API on this is slightly inconsistent and confusing. I'd have preferred a `name, value` pair where the name is the full DNS name and the value whatever the DNS value for that record is. However, this is what we have now.

lzap replied:
Oh this was added as an accident! Optional arguments are only valid for `ib_find_*` methods, this was not intended.

ekohl commented on extra/end-to-end-test.rb:
```diff
@@ -0,0 +1,66 @@
+  cname = "alias-end2end.#{DNS_DOMAIN}"
+  puts "Performing DNS tests with #{fqdn}"
+  begin
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}")
```
It's very weird to me that it even accepts these as GET parameters. IMHO the proxy shouldn't accept that.

lzap replied:
Yeah, I haven't noticed but it worked. It's a different story tho.

ekohl commented on extra/end-to-end-test.rb:
```diff
@@ -0,0 +1,66 @@
+  begin
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}")
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}") # skip (already exists)
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=8.8.8.8") rescue RestClient::Conflict # expected
```
Perhaps it should raise an error if there is no conflict? This will silently accept HTTP 200

I'd also not use any real IPs. rfc5735 suggests using TEST-NET-1 (`192.0.2.0/24`), TEST-NET-2 (`198.51.100.0/24`) or TEST-NET-3 (`203.0.113.0/24`) for this.

lzap replied:
That's how our common DNS code was designed, I won't change it from here. If you look to `dns_common.rb` it silently returns (`nil`). Probably something our orchestration code in Foreman core expects as well.

I replaced Google DNS with `192.0.2.1`.

ekohl commented on extra/end-to-end-test.rb:
```diff
@@ -0,0 +1,66 @@
+# subnets and domains must exist on Infoblox
+URL = "http://localhost:8448"
+DNS_DOMAIN = "ibx.lan"
+DNS_IPV4 = "192.168.202.13"
```
I'd also not use any real IPs. rfc5735 suggests using TEST-NET-1 (`192.0.2.0/24`), TEST-NET-2 (`198.51.100.0/24`) or TEST-NET-3 (`203.0.113.0/24`) for this. Even if it's unlikely that something is routing the NAT network, test nets are intended for this.

lzap replied:
Changed, these IPs are real IP from my Infoblox setup, but I've introduced ENV variables to override this in this test script.

lzap replied:
Also changed the domain to "example.com".

ekohl commented on extra/end-to-end-test.rb:
```diff
@@ -0,0 +1,66 @@
+require 'pp'
+
+# subnets and domains must exist on Infoblox
+URL = "http://localhost:8448"
```
What's on 8448?

lzap replied:
Oh that's smart proxy HTTP port on my setup, I have it for a very long time now, is it different by default? Oh I see it's 8000. Changing.

ekohl commented on lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb:
```diff
@@ -5,7 +5,7 @@ class Record < ::Proxy::Dns::Record
     def initialize(host, connection, ttl, dns_view)
       @connection = connection
       @dns_view = dns_view
-      super(ENV['INFOBLOX_DNS_RESOLVER'] || host, ttl)
```
Can you also remove the mention of this var from the example config?

lzap replied:
Right, good idea.

lzap commented:
All done, rebased. If you are good I am going to merge this and cut new releases of both plugins. Multiple bugs were fixed.

ekohl commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ Gemfile.lock
 .ruby-*
 tags
 config/*.yml
+extra/end-to-end-test.sh
```
Not that I mind that much, but this is very specific :)

lzap replied:
Yeah fixed.

ekohl commented on extra/end-to-end-test.rb:
```diff
@@ -0,0 +1,66 @@
+  begin
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}")
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=#{DNS_IPV4}") # skip (already exists)
+    do_post("/dns?type=a&fqdn=#{fqdn}&value=192.0.2.1") rescue RestClient::Conflict # expected
```
My point was that if this doesn't raise an exception, it's masking a problem. There must be a conflict or you're violating the protocol. The current code can not raise an exception and pass. Since this is actually what's being changed, I'd like to see specific verification of this (since I don't have an infoblox instance to test with).

lzap replied:
I am not following, this should (and does) return HTTP 409 conflict, which makes RestClient to throw this exception. Please explain in more detail what you want me to do here.

ekohl replied:
I'm talking about the other case. If it doesn't raise a conflict, it's a problem. So roughly:
```suggestion
    begin
        do_post("/dns?type=a&fqdn=#{fqdn}&value=192.0.2.1")
        raise Exception # We expect a conflict but didn't get one.
    rescue RestClient::Conflict
        # expected
    end
```

lzap replied:
Sure I am back. Done.

lzap commented:
Rebasing:

```diff
diff --git a/.gitignore b/.gitignore
index cf9534d..d76dc6c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -8,4 +8,4 @@ Gemfile.lock
 .ruby-*
 tags
 config/*.yml
-extra/end-to-end-test.sh
+extra/*.local.sh
diff --git a/.rubocop.yml b/.rubocop.yml
index d436b09..fba2733 100644
--- a/.rubocop.yml
+++ b/.rubocop.yml
@@ -4,6 +4,7 @@ AllCops:
   TargetRubyVersion: 2.0
   Exclude:
     - 'extra/**/*.rb'
+    - 'vendor/**/*'
 
 # There's if statement in the Gemfile
 Bundler/DuplicatedGem:
diff --git a/lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb b/lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb
index a8e41c8..2b6dc7b 100644
--- a/lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb
+++ b/lib/smart_proxy_dns_infoblox/dns_infoblox_main.rb
@@ -86,7 +86,7 @@ module Proxy::Dns::Infoblox
       Infoblox::Cname.find(connection, params)
     end
 
-    def ib_create_a_record(fqdn, address = nil)
+    def ib_create_a_record(fqdn, address)
       ib_create(Infoblox::Arecord, :name => fqdn, :ipv4addr => address)
     end
```

ekohl commented:
Could you have a look at the conflict?

lzap commented:
Done!
